cmake/Utils.cmake

```diff
@@ -0,0 +1,14 @@
+# Collect all currently added targets in all subdirectories
+#
+# Parameters:
+# - _result the list containing all found targets
+# - _dir root directory to start looking from
+function(get_all_targets _result _dir)
+    get_property(_subdirs DIRECTORY "${_dir}" PROPERTY SUBDIRECTORIES)
+    foreach(_subdir IN LISTS _subdirs)
+        get_all_targets(${_result} "${_subdir}")
+    endforeach()
+
+    get_directory_property(_sub_targets DIRECTORY "${_dir}" BUILDSYSTEM_TARGETS)
+    list(APPEND ${_result} ${_sub_targets})
+endfunction()
```

Comment:
Part of the problem here is that CMake passes variables by value, not by reference. Any changes made within this function will remain local to the function and go out of scope on return. Unless you are using a macro (please don't call macros recursively), the only way to export results from a CMake function is to use `set(<variable_name> <value> PARENT_SCOPE)`.

Here is one approach to getting the results you intend:

```suggestion
function(get_all_targets _result _dir)
    get_property(_subdirs DIRECTORY "${_dir}" PROPERTY SUBDIRECTORIES)
    set(local_result "")
    foreach(_subdir IN LISTS _subdirs)
        get_all_targets(child_result "${_subdir}")
        list(APPEND local_result ${child_result})
    endforeach()

    get_directory_property(_sub_targets DIRECTORY "${_dir}" BUILDSYSTEM_TARGETS)
    foreach(_tgt IN LISTS _sub_targets)
        # We need to filter targets; many of them are not the targets we are looking for
        # In our case, the property TYPE is set to EXECUTABLE for the targets we want.
        # This may not be true in your case; you may need to modify this check to allow libraries as well.
        get_target_property(_target_type ${_tgt} TYPE)
        if(${_target_type} STREQUAL "EXECUTABLE")
            message(DEBUG "It's an executable target: ${_tgt}")
            list(APPEND local_result ${_tgt})
        else()
            message(DEBUG "Target ${_tgt} is a(n) ${_target_type}, not an executable")
        endif()
    endforeach()
    
    # If we don't export _result to the parent scope, none of our changes will be saved
    set(${_result} ${local_result} PARENT_SCOPE)
endfunction()
```

It is likely that we will want the property type check to use other criteria. TYPE == EXECUTABLE is convenient here, just because our targets are executables and not libraries; it would be better to allow libraries as well. However, we have "catch_main" as one of our targets, and that's a static library that we did not write. I don't think we need static analysis to run on "catch_main".

If you want to look into alternate property types we could use to filter out targets we don't want, [this stackoverflow answer](https://stackoverflow.com/questions/32183975/how-to-print-all-the-properties-of-a-target-in-cmake) has a few approaches to printing all of the properties available for a target.